Fix scalar overload name constructed by ReplaceWithVeclib.cpp

tex3d opened the pull request:
ReplaceWithVeclib.cpp would construct overload name using all the arguments in the intrinsic, but overloads should only be constructed from arguments for which isVectorIntrinsicWithOverloadTypeAtArg returns true, including the return type first (index -1).

Additionally,
- skip when `Intrinsic::not_intrinsic`, otherwise `isVectorIntrinsicWithOverloadTypeAtArg` asserts for some IntrinsicCalls.

Unblocks translation for pow and atan2 intrinsics.

Fixes #111093

tex3d committed:
Fix scalar overload name constructed by ReplaceWithVeclib.cpp

ReplaceWithVeclib.cpp would construct overload name using all the arguments in the intrinsic, but overloads should only be constructed from arguments for which isVectorIntrinsicWithOverloadTypeAtArg returns true, including the return (-1).

Fixes #111093

farzonl commented on llvm/lib/CodeGen/ReplaceWithVeclib.cpp:
```diff
@@ -108,8 +108,22 @@ static bool replaceWithCallToVeclib(const TargetLibraryInfo &TLI,
   for (auto Arg : enumerate(II->args())) {
     auto *ArgTy = Arg.value()->getType();
+    // Gather type if it is used in the overload name.
+    if (isVectorIntrinsicWithOverloadTypeAtArg(IID, Arg.index())) {
```
We seem to be duplicating things used later and i'm wonder if  it wouldn't be better to store some of this stuff in variables like :
- `isVectorIntrinsicWithScalarOpAtArg(IID, Arg.index()`
- `auto *VectorArgTy = dyn_cast<VectorType>(ArgTy)`

If you do that  you should be able to flatten the conditional  and combine it with the conditionals on 127 ans 129

farzonl commented on llvm/lib/CodeGen/ReplaceWithVeclib.cpp:
```diff
@@ -129,7 +143,7 @@ static bool replaceWithCallToVeclib(const TargetLibraryInfo &TLI,
   // using scalar argument types.
   std::string ScalarName =
       Intrinsic::isOverloaded(IID)
-          ? Intrinsic::getName(IID, ScalarArgTypes, II->getModule())
```
I feel weird about this. Does `ScalarArgTypes` get used anywhere else? From the PR perspective now the code that pushes to the `SmallVector` isn't being used, but you have left it in.

tex3d replied:
It's used to construct the function type `ScalarFTy` below.  This is why it needs all the argument types.

mgabka commented:
you need also to adjust more tests:
replace-with-veclib-sleef-scalable.ll
replace-with-veclib-sleef.ll

mgabka commented on llvm/lib/CodeGen/ReplaceWithVeclib.cpp:
```diff
@@ -108,8 +108,22 @@ static bool replaceWithCallToVeclib(const TargetLibraryInfo &TLI,
   for (auto Arg : enumerate(II->args())) {
     auto *ArgTy = Arg.value()->getType();
+    // Gather type if it is used in the overload name.
+    if (isVectorIntrinsicWithOverloadTypeAtArg(IID, Arg.index())) {
```
are there any tests, testing this logic? I think the pow case is working thanks to the condition added to line 114

tex3d replied:
The `pow` case is impacted by both this logic and return type logic, since `OloadTys` should include only the return type for `pow`, not each of the argument types.  `isVectorIntrinsicWithOverloadTypeAtArg()` is how you determine the arguments that the intrinsic is overloaded on, when not `void`, and defaults to -1 (return type).

Other tests will be running through this logic (and turning out the same), but the pow one is the only case with two arguments which had existing tests meant to match a vector lib call right now.  `atan2` will be added soon, which also needs this.  Other intrinsics could require this change in theory, if there's a matching vector lib call for them.

tex3d replied:
There are three tests impacted by this change.  If you think there's a specific deficiency at this point not addressed by existing testing, please let me know.  For now, I'll be resolving this comment.  Please re-activate if you think it still needs addressing.

mgabka replied:
ah, there is one unit tests using "powi", which is overloaded on a scalar type.
For now we do not have mappings for it in the TLI, so there is no lit tests for it.


The regular pow is not affecting this code path as is only overloaded only on return type.

tex3d committed:
Simplify logic a bit, handle scalar return overload type.

tex3d committed:
Skip not_intrinsic; skip void return for OloadTys; update tests

farzonl commented on llvm/lib/CodeGen/ReplaceWithVeclib.cpp:
```diff
@@ -100,27 +100,38 @@ static void replaceWithTLIFunction(IntrinsicInst *II, VFInfo &Info,
-      ScalarArgTypes.push_back(VectorArgTy->getElementType());
+    auto *ScalarArgTy = ArgTy->getScalarType();
+    ScalarArgTypes.push_back(ScalarArgTy);
+    if (isVectorIntrinsicWithOverloadTypeAtArg(IID, Arg.index()))
```
```suggestion
    bool RequiresOverloadType = isVectorIntrinsicWithOverloadTypeAtArg(IID, Arg.index());
    if (RequiresOverloadType)
```

tex3d replied:
I don't understand the need for this.  It's the only place this is called for this argument index.

farzonl commented on llvm/lib/CodeGen/ReplaceWithVeclib.cpp:
```diff
@@ -100,27 +100,38 @@ static void replaceWithTLIFunction(IntrinsicInst *II, VFInfo &Info,
       else if (EC != VectorArgTy->getElementCount())
         return false;
-    } else
+    } else if (!isVectorIntrinsicWithScalarOpAtArg(IID, Arg.index()))
```
```suggestion
    }
    if (!RequiresOverloadType)
```

tex3d replied:
This is a different call for a different property: `ScalarOpAtArg`, not `OverloadTypeAtArg`.

farzonl replied:
if `isVectorIntrinsicWithOverloadTypeAtArg(IID, Arg.index());` is false on line 125 then it will be false here too. `IID` and `Arg.index()` do not change in the 9 lines in between.

farzonl replied:
Anyways we can ignore this comment if we go with https://github.com/llvm/llvm-project/pull/111095/files#r1790956438

tex3d replied:
These two functions are not the same:
`isVectorIntrinsicWithScalarOpAtArg`
`isVectorIntrinsicWithOverloadTypeAtArg`

farzonl commented on llvm/lib/CodeGen/ReplaceWithVeclib.cpp:
```diff
@@ -100,27 +100,38 @@ static void replaceWithTLIFunction(IntrinsicInst *II, VFInfo &Info,
+    ScalarArgTypes.push_back(ScalarArgTy);
+    if (isVectorIntrinsicWithOverloadTypeAtArg(IID, Arg.index()))
+        OloadTys.push_back(ScalarArgTy);
+    if (auto *VectorArgTy = dyn_cast<VectorType>(ArgTy)) {
```
So we changed the behavior here  before it was  `isVectorIntrinsicWithScalarOpAtArg(IID, Arg.index())` is true and `VectorArgTy` is not NULL. Now you changed it that `isVectorIntrinsicWithScalarOpAtArg` can be true or false and we still  run this block. Why is this the case?

Does it make sens to do 
```c++
if (!isVectorIntrinsicWithScalarOpAtArg(IID, Arg.index()))
   return false;
OloadTys.push_back(ScalarArgTy);
if (auto *VectorArgTy = dyn_cast<VectorType>(ArgTy)) {
...
}
```

tex3d replied:
> before it was isVectorIntrinsicWithScalarOpAtArg(IID, Arg.index()) is true and VectorArgTy is not NULL

That's not quite right.  It was if `isVectorIntrinsicWithScalarOpAtArg` is false and VectorArgTy is not NULL.

The difference is that before, if `isVectorIntrinsicWithScalarOpAtArg` returned true and you had a vector arg type, we would have added the vector arg type instead of a scalar type to the overload and the function signature.  This would be a vector arg present in the scalar form of the intrinsic that doesn't scale with vectorization of the intrinsic.

I don't think the trivially vectorizable intrinsic path is meant to handle that case in the first place, so I didn't think this code is meant to do so either.  So the assumption was made that we only need to add the scalar types to these sets, which simplified the code a bit.

If you think we do want to handle that case, I can change the code back to adding the (potentially) vector argument type when `isVectorIntrinsicWithScalarOpAtArg` is true.

I can think of a sort of intrinsic that could do that: a resource load or store intrinsic with a vector for coordinates that doesn't match the vector dimension for the width of the load or store.  Would we want these to be considered trivially vectorizable intrinsics?

tex3d replied:
In any case, we definitely should not do this up front:
```
if (!isVectorIntrinsicWithScalarOpAtArg(IID, Arg.index()))
   return false;
```

Otherwise, any intrinsic with a vector operand will be skipped.  That's not what this code does.  It expects a vector or scalar operand.  If vector, the scalar type is used and it checks that the size matches, otherwise, it's a scalar operand, this scalar type is used, and `isVectorIntrinsicWithScalarOpAtArg` should return true (it would return false for vector operands).

tex3d replied:
I've updated the change to preserve vectors when considered `isVectorIntrinsicWithScalarOpAtArg` just to be on the safe side.  The latest change returns the flow control in the loop to something closer to the original.

tex3d committed:
Rework logic to preserve "scalar" operand/return type, even if vector

If isVectorIntrinsicWithScalarOpAtArg() returns true for an argument or return, we should not attempt to modify the type or use the element count, even if it is a vector.  This allows the op to be identified as something that doesn't vectorize/scalarize with the width of the intrinsic, but remains the same, even if it is a vector argument.

Initial ElementCount will only be the return element count if it's a vector and isVectorIntrinsicWithScalarOpAtArg returns false.

This also returns the control flow in the loop to be closer to the original.

Remove inconsistent comment about VFABI return type assumption, since we have the accessors that tell us how to handle it for the operation.

Fix formatting.

tex3d committed:
fix formatting

mgabka commented on llvm/lib/CodeGen/ReplaceWithVeclib.cpp:
```diff
@@ -100,20 +100,39 @@ static void replaceWithTLIFunction(IntrinsicInst *II, VFInfo &Info,
 static bool replaceWithCallToVeclib(const TargetLibraryInfo &TLI,
                                     IntrinsicInst *II) {
   assert(II != nullptr && "Intrinsic cannot be null");
-  // At the moment VFABI assumes the return type is always widened unless it is
```
this comment was useful as it explains the why EC is computed in such a way please do not delete it

tex3d replied:
I made the change to rely on using `isVectorIntrinsicWithScalarOpAtArg` as well because it seemed like the right thing to do.  If it ever returned true for the return index `-1`, that would contradict the comment.

However, since I see other paths do not use this for the return and this doesn't return true for index `-1` for any intrinsic, I can change this part back, since it would likely require other code changes if this assumption were to ever change.

mgabka commented on llvm/lib/CodeGen/ReplaceWithVeclib.cpp:
```diff
@@ -100,20 +100,39 @@ static void replaceWithTLIFunction(IntrinsicInst *II, VFInfo &Info,
   SmallVector<Type *, 8> ScalarArgTypes;
-  Intrinsic::ID IID = II->getIntrinsicID();
+  auto *VTy = dyn_cast<VectorType>(RetTy);
+  ElementCount EC(!RetIsScalar && VTy ? VTy->getElementCount()
```
the extra !RetIsScalar is not needed, if the return type is vector then isVectorIntrinsicWithScalarOpAtArg(IID, -1) return false, the original place where EC was computed was correct, could you please revert this change?

mgabka commented on llvm/lib/CodeGen/ReplaceWithVeclib.cpp:
```diff
@@ -100,20 +100,39 @@ static void replaceWithTLIFunction(IntrinsicInst *II, VFInfo &Info,
+    return false;
+
+  // RetIsScalar: Return type is not widened.
+  bool RetIsScalar = isVectorIntrinsicWithScalarOpAtArg(IID, -1);
```
I believe that VFABI does not allow such cases so in my opinion such cases should not be handled,  but simply rejected, at the very beginning of this function.

tex3d commented on llvm/lib/CodeGen/ReplaceWithVeclib.cpp:
```diff
@@ -100,20 +100,39 @@ static void replaceWithTLIFunction(IntrinsicInst *II, VFInfo &Info,
-  ElementCount EC(VTy ? VTy->getElementCount() : ElementCount::getFixed(0));
+  Intrinsic::ID IID = II->getIntrinsicID();
+  if (IID == Intrinsic::not_intrinsic)
+    return false;
```
It only calls this for IntrinsicInst, but some of those have `getIntrinsicID() == Intrinsic::not_intrinsic`.  This caused the assert to fire in `isVectorIntrinsicWithOverloadTypeAtArg`.  Some hits were in an AMDGPU nop intrinsic, but there were others as well.  I'm afraid I don't still have the list of tests on hand that asserted without this.

Before this check, we would have exited a little later, usually when an argument returned false from `isVectorIntrinsicWithScalarOpAtArg` and also wasn't a vector type.

mgabka replied:
Ah I see, the casting method uses  just  ```bool isIntrinsic() const``` which only checks if the name starts with the llvm. and is not checking if there is intrinsic ID, while there seems to be tests using llvm. functions which aren't intrinsics.

That is a but in the top level function in this pass, could you move this extra check to runImpl function so we have all the filtering in one place?

tex3d replied:
Updated.

tex3d committed:
Revert change to use isVectorIntrinsicWithScalarOpAtArg for return type

tex3d committed:
Move intrinsic ID check to loop

tex3d commented:
Note: buildkite failure is an ongoing unrelated failure with:
```
RuntimeError: __cxx03/__algorithm/adjacent_find.h: Header cxx03 is not a valid header
```
